rnnalyse/extractors/base_extractor.py

```diff
@@ -147,38 +145,76 @@ def _print_time_info(self, start_time: float, print_every: int) -> None:
+        num_extracted: int
+            Number of extracted activations for this sentence.
         """
+
+        sen_activations: PartialActivationDict = self._init_sen_activations(0)
+
+        activations: FullActivationDict = self.init_lstm_states.states
+
+        for i, token in enumerate(sentence.sen):
+            out, activations = self.model(token, activations)
+
+            # Check whether current activations match criterion defined in selection_func
+            if self.selection_func(i, token, sentence):
+                for l, name in self.activation_names:
+                    sen_activations[(l, name)] = np.append(
+                        sen_activations[(l, name)], activations[l][name].detach().numpy()[np.newaxis, ...], axis=0
+                    )
+
+        num_extracted = list(sen_activations.values())[0].shape[0]
+        self._dump_activations(sen_activations)
+
+        return num_extracted
+
+    def extract_average_eos_activations(self):
+        """ Extract average end of sentence activations and dump them to a file. """
+
         def _incremental_avg(old_avg, new_value):
             # n_sens only get incremented after _extract_sentence is called, therefore + 1 here
             return old_avg + 1 / (self.n_sens + 1) * (new_value - old_avg)
 
-        sen_activations: PartialActivationDict = self._init_sen_activations(len(sentence))
+        # Init states
+        avg_eos_states = {
+            l: {'hx': torch.zeros(self.model.hidden_size), 'cx': torch.zeros(self.model.hidden_size)}
+            for l in range(self.model.num_layers)
+        }
 
-        activations: FullActivationDict = self.init_lstm_states.states
+        # Extract
+        for sentence in self.corpus.values():
+            sen_activations: PartialActivationDict = self._init_sen_activations(len(sentence))
 
-        for i, token in enumerate(sentence):
-            out, activations = self.model(token, activations)
+            activations: FullActivationDict = self.init_lstm_states.states
+
+            for i, token in enumerate(sentence.sen):
+                if self.selection_func(i, token, sentence):
+                    _, activations = self.model(token, activations)
 
-            for l, name in self.activation_names:
-                sen_activations[(l, name)][i] = activations[l][name].detach().numpy()
+                    for l, name in self.activation_names:
+                        sen_activations[(l, name)] = activations[l][name].detach().numpy()
 
-        # Update average eos states
-        self.avg_eos_states = {
-            l: {
-                'hx': _incremental_avg(self.avg_eos_states[l]['hx'], activations[l]['hx']),
-                'cx': _incremental_avg(self.avg_eos_states[l]['cx'], activations[l]['cx'])
+            # Update average eos states if last activation was extracted
+            avg_eos_states = {
+                l: {
+                    'hx': _incremental_avg(avg_eos_states[l]['hx'], activations[l]['hx']),
```

Comment:
the reason that activations[l]['hx'] contains the last activation for the current sentence is that you don't store the intermediate ones?

Reply:
I would say `activations[l]['hx']` contain the last activations because they're the last value that is assigned to this variable in the last loop of `for i, token in enumerate(sentence.sen)`. I actually just realized that I could omit the whole inner loop in this function but I am just going to reuse the code in `_extract_sentence` like you suggested.